src/lib/completion.ts

```diff
@@ -1,131 +1,223 @@
-import { Document } from 'langchain/document';
-
+//https://js.langchain.com/docs/modules/agents/how_to/custom_llm_agent
 import { ChatOpenAI } from 'langchain/chat_models/openai';
 import { OpenAIEmbeddings } from 'langchain/embeddings/openai';
-import { SystemMessage, BaseMessage, AIMessage, HumanMessage } from 'langchain/schema';
-
-import { ChatHistoryType, type ChatHistory } from '$lib/history';
-import { BytesOutputParser } from 'langchain/schema/output_parser';
-
+import {
+	type AgentAction,
+	type AgentFinish,
+	HumanMessage,
+	type InputValues,
+	type AgentStep,
+	AIMessageChunk
+} from 'langchain/schema';
 import { QdrantClient } from '@qdrant/js-client-rest';
+import { type ChatHistory } from '$lib/history';
+import { formatLogToString } from 'langchain/agents/format_scratchpad/log';
+import { RunnableSequence } from 'langchain/schema/runnable';
+import { AgentExecutor } from 'langchain/agents';
+import { SYSTEM, HISTORY, SUFFIX } from './prompt';
+import { getTools } from './tools';
+import { colors } from './colors';
+import fs from 'fs';
+import { OpenAI } from 'langchain/llms/openai';
 
 const DEFAULT_MODEL = 'gpt-3.5-turbo';
 const DEFAULT_COLLECTION = 'default';
-
 export class ChatbotCompletion {
-	private model: ChatOpenAI;
 	private embeddings_model: OpenAIEmbeddings;
-
+	private executor: AgentExecutor | undefined;
+	private openai_api_key: string;
+	private model_name;
+	private verbose: boolean;
 	private qdrant_client: QdrantClient;
-	private qdrant_collection: string;
+	private collection_name: string;
+	private history: ChatHistory[];
+	private do_history: boolean;
```

Comment:
Could this name be changed to better clarify what it does. You should maybe also add a comment explaining what this means as well.